extensions/amp-ad/0.1/amp-ad-api-handler.js

```diff
@@ -148,29 +151,39 @@ export class AmpAdApiHandler {
 
   /**
    * Updates the element's dimensions to accommodate the iframe's
-   *    requested dimensions.
+   * requested dimensions. Notifies the window that request the resize
+   * of success or failure.
    * @param {number|undefined} height
    * @param {number|undefined} width
+   * @param {!Window} source
+   * @param {string} origin
    * @private
    */
-  updateSize_(height, width) {
-    const targetOrigin =
-          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
-    this.baseInstance_.attemptChangeSize(height, width, () => {
-      postMessage(
-          this.iframe_,
-          'embed-size-changed',
-          {requestedHeight: height, requestedWidth: width},
-          targetOrigin,
-          this.is3p_);
-    }, () => {
-      postMessage(
-          this.iframe_,
-          'embed-size-denied',
-          {requestedHeight: height, requestedWidth: width},
-          targetOrigin,
-          this.is3p_);
-    });
+  updateSize_(height, width, source, origin) {
+    this.baseInstance_.attemptChangeSize(height, width,
+        () => this.sendEmbedSizeResponse_(
+            true /* success */, width, height, source, origin),
+        () => this.sendEmbedSizeResponse_(
+            false /* success */, width, height, source, origin));
+  }
+
+  /**
+   * Sends a response to the window which requested a resize.
+   * @param {boolean} success
+   * @param {number} requestedWidth
+   * @param {number} requestedHeight
+   * @param {!Window} source
+   * @param {string} origin
+   * @private
+   */
+  sendEmbedSizeResponse_(
+      success, requestedWidth, requestedHeight, source, origin) {
+    postMessageToWindows(
+        this.iframe_,
+        [{win: source, origin}],
```

Comment:
Indeed, #3019 is the issue that this PR pertains to.

Direct communication achieves a few different things:

1. It offers better performance, especially for things like IntersectionObserver emulation, as it avoids adding forwarding. It also makes it easier to handle double-nesting without adding extra complexity.
2. It supports A4A use case where the returned ad is a non-AMP creative. In that case the ad is rendered in a 3P iframe, but it does not have access to the normal window.context. With the current implementation of messaging there will be no extra work required to support IntersectionObserver or resizing in that case.

Note that the nested iframes can only communicate if they know a per-3P-frame unique sentinel value. This sentinel is available within the frame's contextWindow and has to be explicitly passed to any child iframes (i.e. it functions as an authorization token).

FYI, the underlying changes to the messaging code happened some time ago in #3327 along with the changes for IntersectionObserver and for page visibility in #4021.